src/gui/Icons.cpp

```diff
@@ -47,45 +65,102 @@ QString Icons::trayIconAppearance() const
     return iconAppearance;
 }
 
-QIcon Icons::trayIcon()
+QIcon Icons::trayIcon(QString style)
 {
-    return trayIconUnlocked();
-}
+    if (style == "unlocked") {
+        style.clear();
+    }
+    if (!style.isEmpty()) {
+        style = "-" + style;
+    }
 
-QIcon Icons::trayIconLocked()
-{
     auto iconApperance = trayIconAppearance();
-
-    if (iconApperance == "monochrome-light") {
-        return icon("keepassxc-monochrome-light-locked", false);
+    if (!iconApperance.startsWith("monochrome")) {
+        return icon(QString("keepassxc%1").arg(style), false);
     }
-    if (iconApperance == "monochrome-dark") {
-        return icon("keepassxc-monochrome-dark-locked", false);
+
+    QIcon i;
+#if defined(Q_OS_MACOS) || defined(Q_OS_WIN)
+    if (osUtils->isStatusBarDark()) {
+        i = icon(QString("keepassxc-monochrome-light%1").arg(style), false);
+    } else {
+        i = icon(QString("keepassxc-monochrome-dark%1").arg(style), false);
     }
-    return icon("keepassxc-locked", false);
+#else
+    i = icon(QString("keepassxc-%1%2").arg(iconApperance, style), false);
+#endif
+#if QT_VERSION >= QT_VERSION_CHECK(5, 6, 0)
+    // Set as mask to allow the operating system to recolour the tray icon. This may look weird
+    // if we failed to detect the status bar background colour correctly, but it is certainly
+    // better than a barely visible icon and even if we did guess correctly, it allows for better
+    // integration should the system's preferred colours not be 100% black or white.
+    i.setIsMask(true);
+#endif
+    return i;
+}
+
+QIcon Icons::trayIconLocked()
+{
+    return trayIcon("locked");
 }
 
 QIcon Icons::trayIconUnlocked()
 {
-    auto iconApperance = trayIconAppearance();
+    return trayIcon("unlocked");
+}
 
-    if (iconApperance == "monochrome-light") {
-        return icon("keepassxc-monochrome-light", false);
-    }
-    if (iconApperance == "monochrome-dark") {
-        return icon("keepassxc-monochrome-dark", false);
-    }
-    return icon("keepassxc", false);
+AdaptiveIconEngine::AdaptiveIconEngine(QIcon baseIcon)
+    : QIconEngine()
+    , m_baseIcon(std::move(baseIcon))
+{
 }
 
-QIcon Icons::icon(const QString& name, bool recolor, const QColor& overrideColor)
+void AdaptiveIconEngine::paint(QPainter* painter, const QRect& rect, QIcon::Mode mode, QIcon::State state)
 {
-    QIcon icon = m_iconCache.value(name);
+#if QT_VERSION >= QT_VERSION_CHECK(5, 6, 0)
+    double dpr = !kpxcApp->testAttribute(Qt::AA_UseHighDpiPixmaps) ? 1.0 : painter->device()->devicePixelRatioF();
+#else
+    double dpr = !kpxcApp->testAttribute(Qt::AA_UseHighDpiPixmaps) ? 1.0 : painter->device()->devicePixelRatio();
+#endif
+    QSize pixmapSize = rect.size() * dpr;
 
-    if (!icon.isNull() && !overrideColor.isValid()) {
-        return icon;
+    painter->save();
+    painter->drawPixmap(rect, m_baseIcon.pixmap(pixmapSize, mode, state));
```

Comment:
I get very ugly icons on HDPI screen on Windows if icon is painted via `drawPixmap` method:
![slika](https://user-images.githubusercontent.com/4377556/105424074-edadf980-5c46-11eb-9cd6-b8f8903d0e3c.png)

Icons are drawn correctly If I change the icon pixelmap to be painted via `QIcon::paint`  method e.g. `m_baseIcon.paint(painter, rect, Qt::AlignCenter, mode, state)`
![slika](https://user-images.githubusercontent.com/4377556/105424330-75940380-5c47-11eb-80da-3bb3ab639cf7.png)